### ISSUE TITLE
feature: add leader election for controllers

### ISSUE THREAD

**Issue by ncdc** [kcp-dev/kcp#2216]
### Feature Description

As a kcp platform operator, I would like to be able to run multiple copies of the kcp process for high availability.

As a kcp platform operator, I need to ensure that multiple copies of the controllers do not fight with each other.

### Proposed Solution

We can enable leader election (optionally) so only 1 copy of each controller is active at a time.

### Want to contribute?

- [ ] I would like to work on this issue.

**Comment by sttts** [comment]
There are probably a thousand implementations. Here is one using client-go https://github.com/kube-bind/kube-bind/blob/main/cmd/konnector/cmd/leaderelection.go

**Comment by ncdc** [comment]
Right, no need to reinvent the wheel, since client-go has it already

**Comment by embik** [comment]
Deployment rollout would run normally, but at the end of the rollout no bootstrapping would have run (which might be vital for the hypothetical v0.21.0 upgrade).

**Comment by xrstf** [comment]
Maybe I'm dense, but why would there be no bootstrap? At _some_ point a new 0.21 Pod _has_ to become leader, no?

**Comment by embik** [comment]
> Maybe I'm dense, but why would there be no bootstrapping? At _some_ point a new 0.21 Pod _has_ to become leader, no?

Bootstrapping only happens once at pod startup. So a 0.21 pod doesn't need to become leader at _some_ point, it needs to be leader _at startup_ if 0.21 bootstrapping is supposed to happen.

Let's walk through an example. We have `kcp-0.20-a` and `kcp-0.20-b` pods running. `kcp-0.20-a` is leader. We trigger a rolling update to 0.21.

1. `kcp-0.21-c` starts up while `kcp-0.20-b` terminates. Our `c` pod will not be able to become leader during startup because `a` is still leader. Bootstrapping will not happen in `c`.
2. `kcp-0.21-c` becomes ready. `kcp-0.21-d` starts up while `kcp-0.20-a` terminates. Since `a` terminates, leader election is up. `c` snatches the leader lock while `d` is still pulling image. `d` does not do bootstrapping because it's not leader.
3. Step 2 can theoretically happen continuously with each `kcp-0.21` pod starting. If a leader election happens because the leader gets replaced, the "established" `kcp-0.21` pod becomes leader, and the new one doesn't run bootstrapping.

Maybe I'm missing something, but I would like to avoid an implementation that depends on "we got lucky and leader election worked in our favour".

**Comment by xrstf** [comment]
Maybe bootstrapping should be done via a dedicated Job that runs once. If both 0.20 and 0.21 pods would be able to function with a bootstrapped 0.21 system, then it should not matter when the bootstrappin happens.

Of course this is maybe just piling on additional complexity, but it sounds managing leader election _and_ bootstrapping in the same process is nightmarish.

**Comment by embik** [comment]
Okay, I've given up on making everything work. #2996 will implement leader elections for the existing (core) controllers in kcp, the bootstrapping situation will be dealt with at a later point.
